lib/ansible/modules/cloud/docker/docker_swarm.py

```diff
@@ -0,0 +1,406 @@
+#!/usr/bin/python
+#
+# Copyright 2016 Red Hat | Ansible
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+
+DOCUMENTATION = '''
+---
+module: docker_swarm
+
+short_description: Manage Swarm cluster.
+
+version_added: "2.6"
+
+description:
+     - Init a new Swarm cluster
+     - Add/Remove nodes or managers to an existing cluster
+options:
+    advertise_addr:
+        description:
+            - Externally reachable address advertised
+                to other nodes. This can either be an address/port combination
+                in the form ``192.168.1.1:4567``, or an interface followed by a
+                port number, like ``eth0:4567``. If the port number is omitted,
+                the port number from the listen address is used. If
+                ``advertise_addr`` is not specified, it will be automatically
+                detected when possible.
+        required: false
+    listen_addr:
+        description:
+            - Listen address used for inter-manager
+                communication, as well as determining the networking interface
+                used for the VXLAN Tunnel Endpoint (VTEP). This can either be
+                an address/port combination in the form ``192.168.1.1:4567``,
+                or an interface followed by a port number, like ``eth0:4567``.
+                If the port number is omitted, the default swarm listening port
+                is used.
+        default: '0.0.0.0:2377'
+        required: false
+    force:
+        description:
+            - Use with state C(init) to force creating a new Swarm, even if already part of one.
+            - Use with state C(leave) to Leave the swarm even if this node is a manager.
+        type: bool
+        default: false
+    state:
```

Comment:
I think the problem will be the same if I create a second module just for the nodes because nodes (and managers) could have different states : present, leave (temporary absent) and remove (definitively absent). So if I resume I will have 3 states for the second module : add/leave/remove.
What is your suggestion ?

Reply:
So there are two concerns:

1. the naming of states (add/remove/leave) which are imperative statements
2. the module is managing 2 different object classes

I know there are other modules that break on both accounts too. Not being a docker swarm expert, I don't know if I can come up with the best solution.

1. Maybe use states **present**/**absent**/**unregistered** ? Or is it an option to have **present**/**absent**, and a separate parameter **enabled**
2. The main question is whether there are additional properties to manage that affect the swarm, or affect the node. That would be a good reason to have it split.

Reply:
BTW the 3 states confused me before, but from the examples it seems **init**/**leave** affects the swarm (no node is provided as argument) and **join**/**remove** affects the node as part of the swarm. But I find those states also confusing, maybe there is an issue in the examples ?

Reply:
__init__ is just to create (initialize) a new swarm environment.
__join/leave/remove__ affects nodes

Reply:
I do agree that the states are actions instead of declared intent which is pretty common in Ansible module land (though I don't think it's a formalized Guideline). I don't think it's a hard blocker for me in the case where the module states are more native to what a docker swarm user would expect.

Reply:
Fine for me. It's not always easy for a reviewer to propose better paradigms, and they do not always exist.

But the legacy of modules doing this wrong is greater than it should be. We ship a lot of bad examples :-)